Spine Yaw Compensation to keep spine trackers from drifting

jabberrock opened the pull request:
Prevents the spine trackers from drifting, by nudging them to stay aligned to the headset.

Players are usually standing facing forwards or walking forwards, where the spine trackers are aligned to the yaw of the headset. However due to yaw drift, the spine trackers will eventually rotate to the side, requiring a reset. This feature adds a small centering rotation to compensate for the yaw drift. (For now, this compensation only activates when your trackers are pointing "up", e.g. when you're standing or sitting straight up.)

The correction amount should be approximately the maximum yaw drift of your gyroscope (0.3 degrees/sec is a good default).

![image](https://github.com/user-attachments/assets/ebc0f770-9d90-4823-a5b2-6ab880332b12)

Eirenliel commented:
Please rebase SolarXR, it was merged.

Eirenliel commented:
I think this can be enabled for legs too since it only works when standing up! Ideally, the setting would be part of Drift compensation setting that's now with 3 selectors: Off, Static/Simple (Old one), Dynamic/Advanced (this one). Then for this to have a setting of compensation speed, maybe a slider in 0.05deg/min intervals, and a selector for which parts to apply it: Upper body, Legs, Arms, Feet. There also needs to be an infomation that Dynamic only works with HMD and Static can work without it.

But this can be merged after review already, and settings can be added in the future. Very excited to RC this!

jabberrock commented:
@Eirenliel thanks taking a look! Definitely can look into putting this into the drift compensation setting later. (Or do you think I should just update it now?)

If this works well, I was going to add alignment between upper leg and lower leg.

But I'm not sure if upper leg can be aligned to the spine. For example, when I stand straight, I notice that my legs are twisted slightly away from my body. I was thinking maybe "balancing" the left and right legs so that they're at the same angle away from the spine...

Eirenliel commented:
> If this works well, I was going to add alignment between upper leg and lower leg. Maybe even foot?

Upper and lower leg should have the same yaw, so they're already averaged.

> But I'm not sure if upper leg can be aligned to the spine. For example, when I stand straight, I notice that my legs are twisted slightly away from my body. I was thinking maybe "balancing" the left and right legs so that they're at the same angle away from the spine...

It shouldn't matter if you reset the same way too. Also since it's a setting people can turn it off if it doesn't work for them and we can improve it in the future.

>  (Or do you think I should just update it now?)

If you want, we're not merging util Monday anyway.

Also pelase fix build errors (lint)

Eirenliel commented:
This needs to be Beta'd before it can be merged.

jabberrock commented:
@Eirenliel How do I get the build system to rebuild, and how does the beta process work? Sorry, it's my first PR. Thanks!

Erimelowo commented:
Since you're a first-time contributor, someone else needs to add their approval for the CI to run.

jabberrock commented:
Sigh oops I didn't realize there was a Java linter too

Eirenliel commented:
For Beta you need to create a post in beta-testing-forum on Discord! If you can't, ask some contributors

Erimelowo commented on server/core/src/main/java/dev/slimevr/config/YawCorrectionConfig.kt:
```diff
@@ -0,0 +1,9 @@
+package dev.slimevr.config
+
+class YawCorrectionConfig {
+	private var _amountInDegPerSec = 0.0f
```
Could also add a toggle to allow simply toggling the option on/off without having to set it to 0. This would also allow the _amountInDegPerSec to have a sensible default value.

Also get/set are automatically generated in Kotlin and you don't need a private member + public propriety (see how other config classes were done)

jabberrock replied:
When JSON deserializes a missing field into the float (e.g. when loading a config from a previous version of the server), it sets it to `NaN`. `@JsonInclude(Include.NON_NULL)` did not seem to work.

Do you know how to fix this?

jabberrock replied:
@Erimelowo could you review https://github.com/SlimeVR/SolarXR-Protocol/pull/155 when you have a chance? Thanks!

jabberrock replied:
Addressed by adding enabled setting.

Erimelowo commented on server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.kt:
```diff
@@ -519,6 +524,72 @@ class HumanSkeleton(
+		// is too small, the gyroscope will overpower the correction and the skeleton
+		// will drift. If it is too big, the player will notice that the skeleton is
+		// rotating when the player doesn't face forward for a long time.
+		val serverUpdateFreqInHz = 1000.0f; // FIXME: Use actual server update rate
```
VRServer has `fpsTimer`

jabberrock replied:
Addressed by using `fpsTimer`

Erimelowo commented on server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.kt:
```diff
@@ -519,6 +524,72 @@ class HumanSkeleton(
+				parentTracker = tracker;
+				continue;
+			}
+
```
Check if the tracker `isImu`. If it's not, don't adjust for drift.

jabberrock replied:
Addressed by adding `isImu` check before updating the yaw. I chose to allow an IMU tracker to be adjusted to match a non-IMU parent tracker.

ButterscotchV commented on server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.kt:
```diff
@@ -519,6 +523,76 @@ class HumanSkeleton(
 	}
 
+	private fun updateTrackerYawCorrections() {
+		val trackersToAlign = listOf(
+			headTracker,
+			upperChestTracker,
+			chestTracker,
+			waistTracker,
+			hipTracker,
+		)
```
This is a ton of allocations (run once per update, aka ~1000 times per second). Ideally we should avoid recreating this, but at the least it should be an array. Perhaps this array should be created in `setTrackersFromList()` and cached for use here?
```suggestion
		val trackersToAlign = arrayOf(
			headTracker,
			upperChestTracker,
			chestTracker,
			waistTracker,
			hipTracker,
		)
```

jabberrock replied:
Addressed by moving it into the `init` of `SpineYawCorrection` class. I also cleaned it up by removing null and non-IMU trackers. This is updated only when the config is changed from the UI.

ButterscotchV commented on server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.kt:
```diff
@@ -13,6 +13,7 @@ import dev.slimevr.tracking.trackers.TrackerRole
 import dev.slimevr.tracking.trackers.TrackerUtils.getFirstAvailableTracker
 import dev.slimevr.tracking.trackers.TrackerUtils.getTrackerForSkeleton
 import dev.slimevr.util.ann.VRServerThread
+import io.eiren.math.FloatMath.toRad
```
Generally in the code we use `deg * FastMath.DEG_TO_RAD`, this method is probably fine to use tho, so idk what people's opinion on this is. This method is a lot shorter to call which I like.

jabberrock replied:
Addressed by switching to the constant.

ButterscotchV commented on server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.kt:
```diff
@@ -519,6 +523,76 @@ class HumanSkeleton(
+	}
+
+	private fun isTrackerPointingUp(trackerRot: Quaternion, maxDeviationInRad: Float): Boolean {
+		val trackerUp = trackerRot.sandwich(POS_Y)
+		return trackerUp.angleTo(POS_Y) < maxDeviationInRad
```
Surely we have a cleaner way of doing this? It doesn't necessarily matter, but I figure I'd open a discussion on this.

jabberrock replied:
This function's going away in the next change. What would be a cleaner way?

ButterscotchV replied:
tbh this is fine

ButterscotchV commented on server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt:
```diff
@@ -327,6 +330,8 @@ class Tracker @JvmOverloads constructor(
 			filteringHandler.getTrackedRotation()
 		}
 
+		rot = EulerAngles(EulerOrder.YZX, 0.0f, yawCorrectionInRad, 0.0f).toQuaternion() * rot
```
`EulerAngles(EulerOrder.YZX, 0.0f, yawCorrectionInRad, 0.0f).toQuaternion()` should be cached on changes to `yawCorrectionInRad`.

jabberrock replied:
Also, this needs to be applied *after* reset and mounting rotations

jabberrock replied:
Addressed by updating a private variable `spineYawCorrectionRot` when `spineYawCorrectionInRad` is set.

jabberrock commented:
@ButterscotchV , I remember you said I should move the `Tracker.yawCorrectionInRad` into `TrackerResetHandler` right?

ButterscotchV commented:
> @ButterscotchV , I remember you said I should move the `Tracker.yawCorrectionInRad` into `TrackerResetHandler` right?

Probably should go there, yeah. The definitions of these things are kinda mangle together, but most of the drift corrective type stuff is around there.
Also maybe some of the code from this PR could be moved *out* of the HumanSkeleton class, as it's largely unrelated besides using the trackers from there?

jabberrock commented:
I realized in SolarXR I named it `yawCorrection`, so I'm going to rename this whole feature "Spine Yaw Correction" for consistency. I updated the names to match.

jabberrock commented:
Additional changes:
1. Extracted algorithm into `SpineYawCorrection` class
2. Moved applying the yaw correction rotation into `TrackerResetsHandler`
3. Renamed all references of "compensation" to "correction"

jabberrock commented:
Freshly tested:
1. When not enabled, no yaw correction happens
2. When enabled, and chest & hip trackers are pointing forward, yaw correction happens
3. When chest tracker is pointing forward, but hip tracker is pointing up, no yaw correction happens

ButterscotchV commented on server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/SpineYawCorrection.kt:
```diff
@@ -0,0 +1,76 @@
+	private val yawCorrectionConfig: YawCorrectionConfig,
+) {
+
+	private val upperBodyTrackers = filterImuTrackers(
+		skeleton.headTracker,
+		skeleton.neckTracker,
+		skeleton.upperChestTracker,
+		skeleton.chestTracker,
+		skeleton.waistTracker,
+		skeleton.hipTracker,
+	)
```
This needs to be updated when the trackers are updated from `HumanSkeleton.setTrackersFromList()`

ButterscotchV replied:
Also does it need to be a list? It feels like an array would suffice still, though it matters less with this usage.

jabberrock replied:
Nice catch! I kept it as a list so that I can use `zipWithNext`.

Addressed by changing `SpineYawCorrection` into a stateful class with `setTrackers` and `setConfig`.

jabberrock commented:
@Erimelowo , could you take another look? Thanks!

@Eirenliel, it looks like the general sentiment of the beta is positive. Most people have been testing v2 and v3, which is a more sophisticated version of this change. Would it be reasonable to merge this?

ButterscotchV reviewed (changes requested):
I've tested with the recommended changes and it seems to work as intended. I had weird results from AutoBone, but I still don't really know whether it's this PR causing it.

ButterscotchV commented on gui/src/components/settings/pages/GeneralSettings.tsx:
```diff
@@ -184,6 +193,10 @@ export function GeneralSettings() {
     unitDisplay: 'narrow',
     maximumFractionDigits: 2,
   });
+  const degreeFormat = new Intl.NumberFormat(currentLocales, {
+    style: 'unit',
+    unit: 'degree',
+  });
```
We should maybe use our translation framework? Idk about this

ButterscotchV replied:
@ImUrX Can I get your feedback on this 🐱

ImUrX replied:
its fine, i discovered a better integrated way on Fluent recently but for now this is how we been doing it, and if its per second you can do `degree-per-second` for the `unit`

ButterscotchV commented on gui/src/components/settings/pages/GeneralSettings.tsx:
```diff
@@ -813,6 +836,55 @@ export function GeneralSettings() {
+                <NumberSelector
+                  control={control}
+                  name="yawCorrectionSettings.amountInDegPerSec"
+                  valueLabelFormat={(value) => degreeFormat.format(value)}
```
This label should say in degrees per second, not just degrees. It is confusing as it is.

ButterscotchV replied:
See https://github.com/SlimeVR/SlimeVR-Server/pull/1243#discussion_r1866370498

ButterscotchV commented on server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/SpineYawCorrection.kt:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	companion object {
+		private fun filterImuTrackers(vararg trackers: Tracker?): List<Tracker> = trackers.filterNotNull().filter { it.isImu() }.toList()
```
Actually, this probably shouldn't filter null trackers either? I'm not sure you would want to correct a hip tracker with the head tracker when every other tracker is missing. If this is intended behaviour then that is fine.

jabberrock commented:
Slightly updated version coming soon. Will create a new PR.
